### PR TITLE
Add `viewer.fit_to_view()`, properly calculate 3D data canvas size, and refactor `viewer.reset_view()`

### PR DESCRIPTION
**Pull request by TimMonko**
# References and relevant issues

A follow-up to #6730 to prepare for implementation of export figure in 3D. 

# Description

1. Adds `viewer.fit_to_view()` which now functions as `viewer.reset_view(reset_camera_angle=False)`. Fit to view does not reset camera angles, only pans and zooms. 
3. Correctly calculates extent of rotated layers in 3D camera; now, `viewer.fit_to_view()` will have proper margins encapsulating the 2D bounding box of this rotated object. (see test script and images below)
4. Refactors `viewer.reset_view()`/`viewer.fit_to_view()` because this method has been growing in use cases and branch points. Adds thorough commenting because I've both had to think hard about the current implementation, and the newer implementations add more burder of understanding. Trying to help future onboarding for this code and increase maintainability
5. Adds unit tests for `reset_view()` and `fit_to_view()` because this was only indirectly covered by tests previously (AFAIK). Adding these tests helped clean up implementation for edge cases, such as 2D-only data rotated in 3D (the vector was length 2 instead of 3, so needed adjustment). 

Test script for visualization comparison
```python
import napari

viewer = napari.Viewer(ndisplay=3)
viewer.open_sample('napari', 'cells3d')
viewer.open_sample('napari', 'cells3d')

viewer.grid.enabled = True
viewer.grid.spacing = 0.1
viewer.camera.angles = (40, 20, 80)
# viewer.reset_view(reset_camera_angle=False) # on main
viewer.fit_to_view() # this PR
napari.run()
```
This PR:
image

Current main:
image

# Next Steps

1. Add `viewer.fit_to_view` to GUI, likely as right click on the 'home' button
2. Add 3D support to `viewer.export_figure()` and related functions
3. Consider changing default `viewer.reset_view()` implementations in various actions to be `viewer.fit_to_view()` by default. I know there are some examples of this, just not looking for it at the moment. See #7765 
4. Improve support for re-order dims and transpose. See https://github.com/napari/napari/pull/7742#issuecomment-2774134537 for more info

### CODE REVIEW

**Comment by TimMonko** [comment]
@aganders3 thanks especially for your support in zulip, [#general > 2D height/width of rotated 3D](https://napari.zulipchat.com/#narrow/channel/212875-general/topic/2D.20height.2Fwidth.20of.20rotated.203D). apologies it took so long -- it's been nagging at me a while, please don't ever feel like helping me out will be forgotten. 😅

File: napari/components/_tests/test_viewer_model.py [hunk 1]
```diff
@@ -1067,3 +1067,114 @@ def test_get_status_text():
...
+    large_margin_zoom = viewer.camera.zoom
+
+    viewer.reset_view(margin=0)
+    no_margin_zoom = viewer.camera.zoom
+
+    # Check zoom decreases with increased margin
+    assert default_zoom > large_margin_zoom
+    assert no_margin_zoom > default_zoom
+
+    # check validation
```
Comment (willingc) [inline napari/components/_tests/test_viewer_model.py]:
```suggestion
    # check margins outside of the supported values
```

File: napari/components/_tests/test_viewer_model.py [hunk 2]
```diff
@@ -1067,3 +1067,114 @@ def test_get_status_text():
...
+    default_zoom = viewer.camera.zoom
+
+    viewer.reset_view(margin=0.2)
+    large_margin_zoom = viewer.camera.zoom
+
+    viewer.reset_view(margin=0)
+    no_margin_zoom = viewer.camera.zoom
+
+    # Check zoom decreases with increased margin
+    assert default_zoom > large_margin_zoom
```
Comment (willingc) [inline napari/components/_tests/test_viewer_model.py]: Style thought: I would normally put the assertion right after the code that sets large_margin_zoom and  no_margin_zoom.

File: napari/components/_tests/test_viewer_model.py [hunk 3]
```diff
@@ -1067,3 +1067,114 @@ def test_get_status_text():
...
+    with pytest.raises(ValueError, match='margin must be between 0 and 1'):
+        viewer.reset_view(margin=1.0)
+
+
+@pytest.mark.parametrize(
+    ('ndisplay', 'expected_center'),
+    [(2, (0, 14.5, 9.5)), (3, (4.5, 14.5, 9.5))],
+)
+def test_reset_view_center_calculation(ndisplay, expected_center):
+    """Test correct center calculation for different dimensions."""
```
Comment (willingc) [inline napari/components/_tests/test_viewer_model.py]:
```suggestion
    """Test correct center calculation for different dimensions after a viewer reset."""
```

File: napari/components/_tests/test_viewer_model.py [hunk 4]
```diff
@@ -1067,3 +1067,114 @@ def test_get_status_text():
...
+    viewer.camera.center = (0, 0, 0)
+    viewer.reset_view()
+
+    # Center should be in the middle of the data, but first coordinate depends on ndisplay
+    np.testing.assert_allclose(viewer.camera.center, expected_center)
+
+
+@pytest.mark.parametrize('reset_angle', [True, False])
+def test_reset_view_angle_reset(reset_angle):
+    """Test camera angle reset behavior."""
```
Comment (willingc) [inline napari/components/_tests/test_viewer_model.py]:
```suggestion
    """Test camera angle behavior after a viewer reset."""
```

File: napari/components/_tests/test_viewer_model.py [hunk 5]
```diff
@@ -1067,3 +1067,114 @@ def test_get_status_text():
...
+
+    # space grid apart, then reset view
+    viewer.grid.spacing = 0.2
+    viewer.reset_view()
+    spaced_grid_zoom = viewer.camera.zoom
+
+    assert spaced_grid_zoom < grid_zoom
+
+
+def test_reset_view_empty():
```
Comment (willingc) [inline napari/components/_tests/test_viewer_model.py]:
```suggestion
def test_reset_view_handles_no_layers():
```

File: napari/components/viewer_model.py [hunk 6]
```diff
@@ -387,13 +387,67 @@ def reset_view(
     ) -> None:
         """Reset the camera view.
 
+        Ajusts the camera zoom and center the view so that all layers are
```
Comment (willingc) [inline napari/components/viewer_model.py]:
```suggestion
        Adjusts the camera zoom and center the view so that all layers are
```

File: napari/components/viewer_model.py [hunk 7]
```diff
@@ -387,13 +387,67 @@ def reset_view(
...
+        visible, accounting for the current grid mode and margin.
+
         Parameters
         ----------
         margin : float in [0, 1)
             Margin as fraction of the canvas, showing blank space around the
-            data.
+            data. Default is 0.05 (5% of the canvas).
+        reset_camera_angle : bool
```
Comment (willingc) [inline napari/components/viewer_model.py:399]:
```suggestion
        use_default_camera_angle : bool
```

File: napari/components/viewer_model.py [hunk 8]
```diff
@@ -421,39 +481,92 @@ def reset_view(
...
         )
         assert len(center) in (2, 3)
-        self.camera.center = center
-        # zoom is defined as the number of canvas pixels per world pixel
-        # The default value used below will zoom such that the whole field
-        # of view will occupy 95% of the canvas on the most filled axis
+        return center
 
+    def _get_scale_factor(self, margin: float) -> float:
+        """Get the scale factor for camera zoom based on a valid margin."""
```
Comment (willingc) [inline napari/components/viewer_model.py]:
```suggestion
        """Get the scale factor for camera zoom with a valid margin."""
```

File: napari/components/viewer_model.py [hunk 9]
```diff
@@ -421,39 +481,92 @@ def reset_view(
...
+        # required for dot product
+        size = np.squeeze(np.diff(extent, axis=0))
+
+        # if the size vector is (2,) and the camera vector is (3,)
+        # add a very small thickness to the size vector in the Z position
+        # to make sure the cross product is valid, and no division by zero
+        if len(size) < len(view_direction):
+            size = np.insert(size, 0, 1e-10)
+
+        # get the "right" direction that is perpendicular to the view and up directions
```
Comment (willingc) [inline napari/components/viewer_model.py]: right or correct?

**Review by willingc — APPROVED** [review]
This looks very good @TimMonko. I've made some style and naming suggestions. Though this is good to :ship: as is too.

File: napari/components/viewer_model.py (hunk 7 above)
Comment (TimMonko) [reply napari/components/viewer_model.py:399]: Are you proposing we rename the argument?

File: napari/components/viewer_model.py (hunk 9 above)
Comment (TimMonko) [reply napari/components/viewer_model.py]: The best way I can explain it, which may be incorrect, is that the view direction is the camera's view--which is rotated around the canvas. As such, if we know the up direction of the canvas, then the right direction (horizontal / Xaxis) would be perpendicular to the view (which is the 'Z' axis we see) and up (the vertical, or Y-axis) directions. As such, by 'right' I'm (well @aganders3) referring to the 'horizontal' component of the canvas. I think it's right direction because its the vector pointing to the right in this perspective. So calculating the dot product of the rotated angle vector and the size vector results in the actual length of the component vector in the 2D 'view direction' space. 

The [docs has a better explanation](https://napari.org/stable/guides/3D_interactivity.html), but maybe I should consider wrapping my mind around this a bit and making a PR to add some of this language to the docs. The last time I did geometry by any measure was 9th grade.

File: napari/components/viewer_model.py (hunk 7 above)
Comment (willingc) [reply napari/components/viewer_model.py:399]: Yeah, though to be honest, I had a hard time finding a name that spoke to me as a bool. reset_camera_angle made me think that it was a value of an angle. Not a big deal and I wouldn't spend more than a few minutes on it.

**Comment by willingc** [comment]
I love the addition of "rightward".

**Comment by jni** [comment]
@TimMonko can you update the issue link in the OP? It's currently pointing to #6370 which seems extremely *un*related to this PR 😂

**Comment by TimMonko** [comment]
good catch. thanks @jni

File: napari/components/viewer_model.py (hunk 7 above)
Comment (TimMonko) [reply napari/components/viewer_model.py:399]: It would make more sense to change it to `reset_camera_angles` in the plural, because its changing multiple angles and its otherwise accessed with `viewer.camera.angles`. But, doesn't changing arg names cause problems for people downstream?

Comment (aganders3) [reply napari/components/viewer_model.py:399]: What do you think of avoiding the boolean arg, and making a new function `def zoom_to_fit(self, margin: float = 0.05) -> None:` and the definition of `reset_view` calls it? Maybe this doesn't work because of ordering? Also I guess `zoom_to_fit` is not quite an accurate name because it may also pan the camera.

```
    def reset_view(self, *, margin: float = 0.05) -> None:
        """Reset the camera view.

        Adjusts the camera zoom and center the view so that all layers are
        visible, accounting for the current grid mode and margin.

        Parameters
        ----------
        margin : float in [0, 1)
            Margin as fraction of the canvas, showing blank space around the
            data. Default is 0.05 (5% of the canvas).
        """
        self.camera.angles = (0, 0, 90)
        self.zoom_to_fit(margin=margin)
```

Comment (TimMonko) [reply napari/components/viewer_model.py:399]: @aganders3 thought about this a while and really dig it. I think creating another 'verb' is much better and we can match the grammar by calling it `fit_to_view`. I think then that makes the two functions very clear 
1. reset_view literally resets
2. fit_to_view will pan/zoom/whatever to make your data fit in the viewer

I'll make a separate PR for this soon because it will also require updating some other objects throughout the codebase and I don't want to lose the focus of this PR.

**Comment by TimMonko** [comment]
This PR now uses @aganders3 suggestion to create a new method (`viewer.fit_to_view()`) and remove the boolean from `viewer.reset_view()`. If we feel this needs iterated on I can remove it from this PR, but there hadn't been much motion on it in the past few days, and I couldn't figure out how to make the new PR depend on this PR _and_ have it show the commits compared to this PR, instead of main. It seems the only way to compare was to my branch, but that is in my repo instead of napari/napari.

File: napari/components/viewer_model.py [hunk 10]
```diff
@@ -383,17 +383,100 @@ def _sliced_extent_world_augmented(self) -> np.ndarray:
...
+                Continue to use `reset_view()` to reset the camera angles.
+                Use `fit_to_view()` to adjust the camera without
+                resetting the camera angles.
         """
+        if reset_camera_angle is not None:
+            warnings.warn(
+                'reset_camera_angle is deprecated since napari 0.6.0 and will be removed in 0.7.0.'
+                'Continue to use `viewer.reset_view()` to reset the camera angles.'
+                'Use `viewer.fit_to_view()` to adjust the camera without resetting the camera angles.',
+                category=DeprecationWarning,
```
Comment (Czaki) [inline napari/components/viewer_model.py]:
```suggestion
                category=FutureWarning,
```
We use `FutureWarning` instead of `DeprecationWarning` as `DeprecationWarning` is silenced by default by python.

**Comment by aganders3** [comment]
Ah sorry, I didn't realize this would be a deprecation. I was reading the diff wrong and didn't realize `reset_camera_angle` was on `main`. I'm still +1 on this change though.

File: napari/components/viewer_model.py (hunk 7 above)
Comment (Czaki) [reply napari/components/viewer_model.py:399]: As it is deprecated, I suggest not to touch this name. Old code will work. New code will use `fit_to_view` or `reset_view` depending on need

**Review by Czaki — APPROVED** [review]
Looks good to me

**Comment by TimMonko** [comment]
Anyone know what is wrong with the `.. deprecated` docstring? CircleCI failed because of incorrect indentation, but it looks ok to me

File: napari/components/viewer_model.py [hunk 11]
```diff
@@ -383,17 +383,100 @@ def _sliced_extent_world_augmented(self) -> np.ndarray:
...
         Parameters
         ----------
         margin : float in [0, 1)
             Margin as fraction of the canvas, showing blank space around the
-            data.
+            data. Default is 0.05 (5% of the canvas).
+        reset_camera_angle : bool
+            Whether to reset the camera angles to (0, 0, 90) before fitting
+            to view. Default is True.
+            .. deprecated:: 0.6.0
```
Comment (jni) [inline napari/components/viewer_model.py]:
```suggestion
            to view. Default is True.

            .. deprecated:: 0.6.0
```

I'm not sure but I think sphinx directives need a blank newline in front of them.

Comment (TimMonko) [reply napari/components/viewer_model.py]: I was going off based of TextManager deprecation format, but that is not passed to docs API build. Instead, I found this formatting which does build in docs correctly:
https://github.com/napari/napari/blob/c36a49d61c348a1c24166391421482da1e3aa2e3/napari/_qt/qt_main_window.py#L1171-L1176

**Comment by TimMonko** [comment]
Also, we definitely don't need to deprecate `reset_camera_angle` in `reset_view()`, but I think from a forward thinking perspective it seems reasonable to separate the grammar and is recent (and niche) enough to probably have few knock on effects.

**Comment by TimMonko** [comment]
It would be lovely if a decision was made on this / this was merged. I have a few other PRs in the works that depend on this, and for my noviceness it would help me read the diffs for it to be on main. 🙃

**Comment by psobolewskiPhD** [comment]
> Add viewer.fit_to_view to GUI, likely as right click on the 'home' button

For now could use this in the View menu (now also in the command palette):
https://github.com/napari/napari/blob/4c9242de028ed18305fc2dae2db32745694f6059/napari/_qt/_qapp_model/qactions/_view.py#L65-L66

**Comment by jni** [comment]
I've restarted the failed builds.

re deprecation — maybe we don't deprecate, since it was added not long ago, so it could be a bit annoying for folks — there is no rush in deprecating it? 🤷

File: napari/_qt/_qapp_model/qactions/_view.py [hunk 12]
```diff
@@ -63,7 +63,7 @@ def _get_current_tooltip_visibility() -> bool:
 
 
 def _fit_to_view(viewer: Viewer):
-    viewer.reset_view(reset_camera_angle=False)
+    viewer.fit_to_view()
```
Comment (psobolewskiPhD) [inline napari/_qt/_qapp_model/qactions/_view.py:66]: Worth it for this alone!
Makes Ctrl/command-0 so satisfying to use!
😍

**Comment by psobolewskiPhD** [comment]
So if I change dims order in grid view, then fit_to_view() doesn't work correctly anymore -- the layers get clipped.
Probably a corner case that's not realistic.
Frankly, grid mode doesn't do what i'd expect with changed dims order.
Like with cells3D "Side view" i'd expect clicking grid to show me the two layers side-by-side. Instead they're front to back.

**Comment by TimMonko** [comment]
> So if I change dims order in grid view, then fit_to_view() doesn't work correctly anymore -- the layers get clipped.
> Probably a corner case that's not realistic.
> Frankly, grid mode doesn't do what i'd expect with changed dims order.
> Like with cells3D "Side view" i'd expect clicking grid to show me the two layers side-by-side. Instead they're front to back.

Both transpose and roll are not working right in these cases, but didn't work correctly before. I tried to work on it for a bit, but wrapping my mind around all the different shapes of arrays present was really not getting me very far. We should open an issue for this. 

I'm also wondering how much effort to put in here, since I think I've seen some intention to rethink ordering/transposing in general.

**Review by willingc — APPROVED** [review]
Let's get this merged and iterate in separate PRs if needed. This PR history is getting long enough that it has a lot of cognitive load to keep reviewing as a large PR. Let's land and work on smaller PRs if there are important things to address still. :smile:

**Review by Czaki — APPROVED** [review]
